e3nn/point/message_passing.py

```diff
@@ -29,3 +31,209 @@ def message(self, x_j, k):
+
+        self.representations = representations
+        self._find_max_l_out_in()
+        self._calculate_coupling_coefficients_and_offsets()
+
+        Rs_in = representations[0]
+        Rs_out = representations[-1]
+
+        radial_model = partial(GaussianRadialModel, max_radius=3.2, min_radius=0.7, number_of_basis=10, h=100, L=3, act=swish)
+
+        self.real_spherical_harmonics = rsh
+        self.layer = TensorPassingLayer(Rs_in, Rs_out, radial_model, self.coupling_coefficients, self.coupling_coefficients_base_offsets, self.max_l_in, self.device)
+
+    def _find_max_l_out_in(self):
+        max_l_out = max_l_in = max_l = 0
+
+        for Rs_in, Rs_out in zip(self.representations[:-1], self.representations[1:]):
+            # running values are need to make sure max_l is not higher that needed (consequent layers)
+            # e.g, rotational orders: [4, 6, 4] -> 12 (without), but [4, 6, 4] -> 10 (with)
+            running_max_l_out = max([l_out for (_, l_out, *_) in Rs_out])
+            running_max_l_in = max([l_in for (_, l_in, *_) in Rs_in])
+
+            max_l_out = max(running_max_l_out, max_l_out)
+            max_l_in = max(running_max_l_in, max_l_in)
+            max_l = max(running_max_l_out + running_max_l_in, max_l)
+
+        assert max_l <= 10, "l > 10 is not supported"
+        self.max_l_out = max_l_out
+        self.max_l_in = max_l_in
+        self.max_l = max_l
+
+    def _calculate_coupling_coefficients_and_offsets(self):
+        from e3nn.o3 import wigner_3j
+
+        tmp_base_offsets_list = []
+        tmp_position = 0
+        for l_out in range(self.max_l_out + 1):
+            for l_in in range(self.max_l_in + 1):
+                tmp_base_offsets_list.append(tmp_position)
+                for l_f in range(abs(l_out - l_in), l_out + l_in + 1):
+                    tmp_position += (2 * l_out + 1) * (2 * l_in + 1) * (2 * l_f + 1)
+
+        self.coupling_coefficients = torch.zeros(tmp_position, device=self.device)      # last tmp_cg_pos (not stored in offsets) = size
+
+        for l_out in range(self.max_l_out + 1):
+            for l_in in range(self.max_l_in + 1):
+                tmp_base_pos = tmp_base_offsets_list[l_out * (self.max_l_in + 1) + l_in]
+                for l_f in range(abs(l_out - l_in), l_out + l_in + 1):
+                    tmp_flat = wigner_3j(l_out, l_in, l_f).contiguous().view(-1).type(torch.get_default_dtype()).to(self.device)
+                    self.coupling_coefficients[tmp_base_pos:tmp_base_pos + tmp_flat.shape[0]] = tmp_flat
+                    tmp_base_pos += tmp_flat.shape[0]
+
+        self.coupling_coefficients_base_offsets = torch.tensor(tmp_base_offsets_list, dtype=torch.int32, device=self.device)
+
+    def forward(self, graph):
+        Y = self.real_spherical_harmonics(list(range(self.max_l + 1)), graph.rel_vec).contiguous()
+        return self.layer(graph.edge_index, graph.x, graph.edge_attr, Y)
+
+
+class TensorPassingLayer(torch_geometric.nn.MessagePassing):
```

Comment:
Please add equivariance tests for this layer -- see `e3nn/tests/point/message_passing_tests.py` for examples.